engine/lite/components/base/src/main/scala/pl/touk/nussknacker/engine/lite/components/Union.scala

```diff
@@ -1,42 +1,49 @@
 package pl.touk.nussknacker.engine.lite.components
 
 import cats.Monad
-import cats.data.Validated.Valid
+import cats.data.Validated
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.{CannotCreateObjectError, NodeId}
 import pl.touk.nussknacker.engine.api.context.{ContextTransformation, JoinContextTransformation, ValidationContext}
-import pl.touk.nussknacker.engine.api.typed.typing.Typed
-import pl.touk.nussknacker.engine.api.{CustomStreamTransformer, MethodToInvoke}
+import pl.touk.nussknacker.engine.api.typed.typing
+import pl.touk.nussknacker.engine.api.{BranchParamName, CustomStreamTransformer, LazyParameter, MethodToInvoke, OutputVariableName}
 import pl.touk.nussknacker.engine.lite.api.commonTypes.{DataBatch, ResultType}
-import pl.touk.nussknacker.engine.lite.api.customComponentTypes.{CustomComponentContext, LiteJoinCustomComponent, JoinDataBatch}
+import pl.touk.nussknacker.engine.lite.api.customComponentTypes.{CustomComponentContext, JoinDataBatch, LiteJoinCustomComponent}
 
 import scala.language.higherKinds
 
 //TODO: unify definition with UnionTransformer
 object Union extends CustomStreamTransformer {
 
   @MethodToInvoke
-  def execute(): JoinContextTransformation = {
+  def execute(@BranchParamName("Output expression") outputExpressionByBranchId: Map[String, LazyParameter[AnyRef]],
+              @OutputVariableName variableName: String)(implicit nodeId: NodeId): JoinContextTransformation = {
     ContextTransformation
       .join
       .definedBy { contexts =>
-        Valid(computeIntersection(contexts))
+        val branchReturnTypes: Iterable[typing.TypingResult] = outputExpressionByBranchId.values.map(_.returnType)
+        ContextTransformation.findUniqueParentContext(contexts).map { parent =>
+          ValidationContext(Map(variableName -> branchReturnTypes.head), Map.empty, parent)
+        }.andThen { vc =>
```

Comment:
It will feel more natural if you check this `branchReturnTypes` before, when computing them (around line 23)

Reply:
done